### ISSUE TITLE
[Story a11y] Inactive page elements are tabbable

### ISSUE THREAD

**Issue by mszylkowski** [ampproject/amphtml#31747]
Items on inactive pages are tabbable, which happens on desktop panels but also mobile.
When tabbing through a story, only the elements on the current story (and system layer) should be tabbable.

This includes links as well as elements in shadow roots, such as disclaimer icon in interactive components.

**List of items that should be disabled for tabbing:**
- [x] Regular CTAs on the page #32840
- [x] Story bookend (story scrolls down to the empty space where it's stored when inactive) #32859
- [x] Share dialog (can show when tabbing on mobile UI) #32859
- [x] Amp-twitter (on inactive pages will scroll the view towards the inactive page) #33919
- [x] Page attachments (attachment covers the current page when focused) #32859

## How to reproduce

On the story at https://www.gstatic.com/amphtml/stamp/qa/tooltip.html press tab until you tab through the next page. If on mobile layout, the page will advance automatically when tabbing to an element on a different page. If on desktop, it will shift the system layer as it focuses on elements outside the viewport.

## Screenshots

Elements outside the active page can be tabbed:
![image](https://user-images.githubusercontent.com/22420856/103236243-605eda00-4912-11eb-8487-6e927cb7bc9c.png)

A problem with this is that tabbing on an item outside the viewport will make the window scroll into position, breaking/shifting the layout (see position of share button):
![image](https://user-images.githubusercontent.com/22420856/103236441-ea0ea780-4912-11eb-8060-9e55acabb613.png)

**Comment by gmajoulet** [comment]
cc @processprocess who did some work on this. Was it for talkback only?

**Comment by processprocess** [comment]
> cc @processprocess who did some work on this. Was it for talkback only?

Yes that work was for TalkBack. It was for focusing content on the active page.
It's related in concept but tabbing works differently.
Tabbing does not respect `aria-hidden`.

Hiding focusable elements from tabbing can be achieved by:

1. Apply tabindex="-1" to the focusable element (which could potentially get complex)
2. Add a hidden attribute to the container element.
3. Apply the display: none or the visibility: hidden CSS property to the container element.

[web.dev/aria-hidden-focus ](https://web.dev/aria-hidden-focus/#how-to-fix-partially-hidden-focusable-elements) for additional context.

Option 2 or 3 could work for mobile and desktop-fullbleed but would break desktop panels in it's current form.

**Comment by mszylkowski** [comment]
Thanks for the context! I think 2 and 3 are not part of the equation since mobile and desktop-fullbleed already works well, this issue just happens on desktop panels. Option 1 seems to be the most useful (and we can apply them independently of the UI type so we don't have to deal with changes on UI type).

As you said option 1 will get complex but could be done looking at the list on [HTML Spec tabindex](https://html.spec.whatwg.org/multipage/interaction.html#the-tabindex-attribute) and the allowed descendants of amp-story-page to handle this on direct page descendants, and handling it on Shadow root elements using custom code on the components (eg the disclaimer/options on interactive components).

**Comment by processprocess** [comment]
> I think 2 and 3 are not part of the equation since mobile and desktop-fullbleed already works well

Which stories are you testing this with?
Tabbing on our manual test story looks like it reveals some bugs that might be related to this.
https://www.gstatic.com/amphtml/stamp/qa/tooltip.html

**Comment by mszylkowski** [comment]
Ohhhh true, that's bad lol! Still option 2 and 3 would break with desktop panels, but the behavior in this issue is 100% present on that story so we could add QA tests when we solve this. Tweets are just as interactive as quizzes, but this **also** breaks with regular links 😅.

**Comment by mszylkowski** [comment]
The amount of elements that are tabbable in the story is much larger than what are here, here are the problematic components:
- Regular CTAs on the page
- Story bookend (story scrolls down to the empty space where it's stored when inactive)
- Share dialog (can show when tabbing on mobile UI)
- Amp-twitter (on inactive pages will scroll the view towards the inactive page)

**Comment by processprocess** [comment]
> The amount of elements that are tabbable in the story is much larger than what are here, here are the problematic components:
> - Regular CTAs on the page
> - Story bookend (story scrolls down to the empty space where it's stored when inactive)
> - Share dialog (can show when tabbing on mobile UI)
> - Amp-twitter (on inactive pages will scroll the view towards the inactive page)

These components are mentioned in https://github.com/ampproject/amphtml/issues/32511
